// @webwaka/core/dist/core/logger/index.js
/**
 * Platform Logger Utility
 *
 * Enforces the "Zero Console Logs" invariant across the WebWaka platform.
 * All logging must go through this module, never direct console.log/warn/error calls.
 *
 * Blueprint Reference: Part 9.3 — "Zero Console Logs: No console.log statements. Must use platform logger."
 */
/**
 * Platform Logger
 *
 * Usage:
 *   logger.info("User logged in", { tenantId: "tenant-123", userId: "user-456" });
 *   logger.warn("API rate limit approaching", { tenantId: "tenant-123" });
 *   logger.error("Database connection failed", { tenantId: "tenant-123" }, error);
 *   logger.debug("Cache hit", { tenantId: "tenant-123", key: "user-profile" });
 */
class PlatformLogger {
    isDevelopment = globalThis.process?.env?.NODE_ENV === 'development';
    /**
     * Log informational message
     */
    info(message, context) {
        this.log('info', message, context);
    }
    /**
     * Log warning message
     */
    warn(message, context) {
        this.log('warn', message, context);
    }
    /**
     * Log error message
     */
    error(message, context, error) {
        let ctx = context;
        let err = error;
        // Handle overload: error(message, error) or error(message, context, error)
        if (context instanceof Error) {
            err = context;
            ctx = undefined;
        }
        this.log('error', message, ctx, err);
    }
    /**
     * Log debug message (development only)
     */
    debug(message, context) {
        if (this.isDevelopment) {
            this.log('debug', message, context);
        }
    }
    /**
     * Internal logging implementation
     */
    log(level, message, context, error) {
        const serializedError = error !== undefined
            ? { message: error.message, ...(error.stack !== undefined ? { stack: error.stack } : {}) }
            : undefined;
        const entry = {
            timestamp: new Date().toISOString(),
            level,
            message,
            ...(context !== undefined ? { context } : {}),
            ...(serializedError !== undefined ? { error: serializedError } : {}),
        };
        // In production, this would send to a logging service (e.g., Datadog, Sentry)
        // For now, we output to stderr to avoid stdout pollution
        if (typeof console !== 'undefined') {
            const output = JSON.stringify(entry);
            switch (level) {
                case 'error':
                    console.error(output);
                    break;
                case 'warn':
                    console.warn(output);
                    break;
                case 'debug':
                    console.debug(output);
                    break;
                case 'info':
                default:
                    console.info(output);
                    break;
            }
        }
    }
}
// Export singleton instance
export const logger = new PlatformLogger();
