// @webwaka/core/dist/ai.js
export class OpenRouterClient {
    apiKey;
    defaultModel;
    baseUrl = 'https://openrouter.ai/api/v1';
    constructor(apiKey, defaultModel = 'openai/gpt-4o-mini') {
        this.apiKey = apiKey;
        this.defaultModel = defaultModel;
    }
    async complete(opts) {
        const model = opts.model ?? this.defaultModel;
        try {
            const body = {
                model,
                messages: opts.messages,
            };
            if (opts.maxTokens !== undefined)
                body.max_tokens = opts.maxTokens;
            if (opts.temperature !== undefined)
                body.temperature = opts.temperature;
            const res = await fetch(`${this.baseUrl}/chat/completions`, {
                method: 'POST',
                headers: {
                    Authorization: `Bearer ${this.apiKey}`,
                    'Content-Type': 'application/json',
                    'HTTP-Referer': 'https://webwaka.com',
                    'X-Title': 'WebWaka Commerce',
                },
                body: JSON.stringify(body),
            });
            const data = (await res.json());
            if (!res.ok) {
                return {
                    content: '',
                    model,
                    tokensUsed: 0,
                    error: data?.error?.message ?? 'OpenRouter request failed',
                };
            }
            return {
                content: data.choices?.[0]?.message?.content ?? '',
                model: data.model ?? model,
                tokensUsed: data.usage?.total_tokens ?? 0,
            };
        }
        catch (err) {
            return { content: '', model, tokensUsed: 0, error: err?.message };
        }
    }
}
export function createAiClient(apiKey, defaultModel) {
    return new OpenRouterClient(apiKey, defaultModel);
}
