// @webwaka/core/dist/payment.js
export class PaystackProvider {
    secretKey;
    baseUrl = 'https://api.paystack.co';
    constructor(secretKey) {
        this.secretKey = secretKey;
    }
    get headers() {
        return {
            Authorization: `Bearer ${this.secretKey}`,
            'Content-Type': 'application/json',
        };
    }
    async verifyCharge(reference) {
        try {
            const res = await fetch(`${this.baseUrl}/transaction/verify/${encodeURIComponent(reference)}`, { method: 'GET', headers: this.headers });
            const data = (await res.json());
            if (!res.ok || !data.status) {
                return {
                    success: false,
                    reference,
                    amountKobo: 0,
                    error: data.message ?? 'Verification failed',
                };
            }
            return {
                success: data.data?.status === 'success',
                reference,
                amountKobo: data.data?.amount ?? 0,
                error: data.data?.status !== 'success' ? data.data?.gateway_response : undefined,
            };
        }
        catch (err) {
            return { success: false, reference, amountKobo: 0, error: err?.message };
        }
    }
    async initiateRefund(reference, amountKobo) {
        try {
            const body = { transaction: reference };
            if (amountKobo !== undefined)
                body.amount = amountKobo;
            const res = await fetch(`${this.baseUrl}/refund`, {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify(body),
            });
            const data = (await res.json());
            if (!res.ok || !data.status) {
                return { success: false, refundId: '', error: data.message ?? 'Refund failed' };
            }
            return { success: true, refundId: String(data.data?.id ?? '') };
        }
        catch (err) {
            return { success: false, refundId: '', error: err?.message };
        }
    }
    async initiateSplit(totalKobo, recipients, reference) {
        const subaccounts = recipients.map((r) => ({
            subaccount: r.subaccountCode,
            amount: r.amountKobo,
        }));
        try {
            const res = await fetch(`${this.baseUrl}/transaction/initialize`, {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify({
                    amount: totalKobo,
                    reference,
                    split: { type: 'flat', subaccounts },
                }),
            });
            const data = (await res.json());
            if (!res.ok || !data.status) {
                return {
                    success: false,
                    reference,
                    amountKobo: totalKobo,
                    error: data.message ?? 'Split initiation failed',
                };
            }
            return { success: true, reference, amountKobo: totalKobo };
        }
        catch (err) {
            return { success: false, reference, amountKobo: totalKobo, error: err?.message };
        }
    }
    async initiateTransfer(recipientCode, amountKobo, reference) {
        try {
            const res = await fetch(`${this.baseUrl}/transfer`, {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify({
                    source: 'balance',
                    amount: amountKobo,
                    recipient: recipientCode,
                    reference,
                }),
            });
            const data = (await res.json());
            if (!res.ok || !data.status) {
                return {
                    success: false,
                    transferCode: '',
                    error: data.message ?? 'Transfer failed',
                };
            }
            return { success: true, transferCode: data.data?.transfer_code ?? '' };
        }
        catch (err) {
            return { success: false, transferCode: '', error: err?.message };
        }
    }
}
export function createPaymentProvider(secretKey) {
    return new PaystackProvider(secretKey);
}
