// @webwaka/core/dist/sms.js
export class TermiiProvider {
    apiKey;
    senderId;
    apiUrl = 'https://api.ng.termii.com/api/sms/send';
    constructor(apiKey, senderId = 'WebWaka') {
        this.apiKey = apiKey;
        this.senderId = senderId;
    }
    async sendOtp(to, message, channel = 'whatsapp') {
        const result = await this.deliver(to, message, channel);
        if (!result.success && channel !== 'sms') {
            return this.deliver(to, message, 'sms');
        }
        return result;
    }
    async sendMessage(to, message) {
        return this.sendOtp(to, message, 'whatsapp');
    }
    async deliver(to, message, channel) {
        const termiiChannel = channel === 'whatsapp' || channel === 'whatsapp_business' ? 'whatsapp' : 'generic';
        try {
            const res = await fetch(this.apiUrl, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    to,
                    from: this.senderId,
                    sms: message,
                    type: 'plain',
                    channel: termiiChannel,
                    api_key: this.apiKey,
                }),
            });
            const data = (await res.json());
            if (!res.ok) {
                return {
                    success: false,
                    channel,
                    error: data?.message ?? 'Termii request failed',
                };
            }
            return {
                success: true,
                messageId: data?.message_id ?? data?.pinId,
                channel,
            };
        }
        catch (err) {
            return { success: false, channel, error: err?.message };
        }
    }
}
export function createSmsProvider(apiKey, senderId) {
    return new TermiiProvider(apiKey, senderId);
}
export async function sendTermiiSms(to, message, apiKey, senderId) {
    const provider = createSmsProvider(apiKey, senderId);
    return provider.sendMessage(to, message);
}
