// @webwaka/core/dist/core/notifications/index.js
/**
 * CORE-7: Unified Notification Service
 * Blueprint Reference: Part 10.12 (Cross-Cutting Functional Modules)
 * Blueprint Reference: Part 9.1 #5 (Nigeria First - Yournotify, Termii)
 *
 * Implements event-driven email, SMS, and push notification dispatchers.
 */
import { logger } from '../logger';
export class NotificationService {
    config;
    constructor(config) {
        this.config = config;
    }
    /**
     * Dispatches a notification based on its type.
     */
    async dispatch(payload) {
        switch (payload.type) {
            case 'email':
                return this.sendEmail(payload);
            case 'sms':
                return this.sendSms(payload);
            case 'push':
                return this.sendPush(payload);
            default:
                throw new Error(`Unsupported notification type: ${payload.type}`);
        }
    }
    /**
     * Sends an email using Yournotify (Nigeria-First Service)
     */
    async sendEmail(payload) {
        if (!this.config.yournotifyApiKey) {
            logger.warn('Yournotify API key missing. Email not sent.', {
                tenantId: payload.tenantId,
                recipient: payload.recipient,
            });
            return false;
        }
        try {
            const response = await fetch('https://api.yournotify.com/v1/campaigns/email', {
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.config.yournotifyApiKey}`,
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    to: payload.recipient,
                    subject: payload.subject || 'Notification from WebWaka',
                    html: payload.body
                })
            });
            if (!response.ok) {
                throw new Error(`Yournotify API error: ${response.statusText}`);
            }
            return true;
        }
        catch (error) {
            logger.error('Failed to send email via Yournotify', { tenantId: payload.tenantId }, error);
            return false;
        }
    }
    /**
     * Sends an SMS using Termii (Nigeria-First Service)
     */
    async sendSms(payload) {
        if (!this.config.termiiApiKey) {
            logger.warn('Termii API key missing. SMS not sent.', {
                tenantId: payload.tenantId,
                recipient: payload.recipient,
            });
            return false;
        }
        try {
            const response = await fetch('https://api.ng.termii.com/api/sms/send', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    to: payload.recipient,
                    from: this.config.termiiSenderId || 'WebWaka',
                    sms: payload.body,
                    type: 'plain',
                    channel: 'generic',
                    api_key: this.config.termiiApiKey
                })
            });
            if (!response.ok) {
                throw new Error(`Termii API error: ${response.statusText}`);
            }
            return true;
        }
        catch (error) {
            logger.error('Failed to send SMS via Termii', { tenantId: payload.tenantId }, error);
            return false;
        }
    }
    /**
     * Sends a push notification (Mock implementation for now)
     */
    async sendPush(payload) {
        logger.info('Push notification sent', {
            tenantId: payload.tenantId,
            recipient: payload.recipient,
            body: payload.body,
        });
        return true;
    }
}
