// @webwaka/core/dist/core/geolocation/index.js
/**
 * CORE-9: Geolocation & Mapping Engine
 * Blueprint Reference: Part 10.3 (Transport), Part 10.4 (Logistics)
 *
 * Vendor-neutral abstraction layer for maps, routing, and geocoding.
 */
export class GeolocationEngine {
    provider;
    constructor(provider = 'osm') {
        this.provider = provider;
    }
    /**
     * Calculates the distance and ETA between two points.
     */
    async calculateRoute(origin, destination) {
        // In a real implementation, this would call the respective provider's API
        // For now, we return a mock calculation based on straight-line distance
        const distance = this.calculateStraightLineDistance(origin, destination);
        // Assume average speed of 30 km/h (8.33 m/s) in city traffic
        const duration = Math.floor(distance / 8.33);
        return {
            distanceMeters: Math.floor(distance),
            durationSeconds: duration,
            polyline: 'mock_polyline_data'
        };
    }
    /**
     * Checks if a coordinate is within a specific geofence (radius in meters).
     */
    isWithinGeofence(point, center, radiusMeters) {
        const distance = this.calculateStraightLineDistance(point, center);
        return distance <= radiusMeters;
    }
    /**
     * Haversine formula to calculate distance between two coordinates in meters.
     */
    calculateStraightLineDistance(coord1, coord2) {
        const R = 6371e3; // Earth's radius in meters
        const φ1 = coord1.lat * Math.PI / 180;
        const φ2 = coord2.lat * Math.PI / 180;
        const Δφ = (coord2.lat - coord1.lat) * Math.PI / 180;
        const Δλ = (coord2.lng - coord1.lng) * Math.PI / 180;
        const a = Math.sin(Δφ / 2) * Math.sin(Δφ / 2) +
            Math.cos(φ1) * Math.cos(φ2) *
                Math.sin(Δλ / 2) * Math.sin(Δλ / 2);
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }
}
