// @webwaka/core/dist/core/billing/index.js
/**
 * CORE-8: Platform Billing & Usage Ledger
 * Blueprint Reference: Part 10.1 (Central Management & Economics)
 * Blueprint Reference: Part 9.1 #6 (Africa First - Integer Kobo Values)
 *
 * Implements internal ledger for tracking tenant API/AI usage and credits.
 */
import { logger } from '../logger';
export var LedgerEntryType;
(function (LedgerEntryType) {
    LedgerEntryType["CREDIT"] = "CREDIT";
    LedgerEntryType["DEBIT"] = "DEBIT";
})(LedgerEntryType || (LedgerEntryType = {}));
export var UsageCategory;
(function (UsageCategory) {
    UsageCategory["AI_TOKENS"] = "AI_TOKENS";
    UsageCategory["SMS_SENT"] = "SMS_SENT";
    UsageCategory["EMAIL_SENT"] = "EMAIL_SENT";
    UsageCategory["SUBSCRIPTION_FEE"] = "SUBSCRIPTION_FEE";
    UsageCategory["SUBSCRIPTION_CREDIT"] = "SUBSCRIPTION_CREDIT";
})(UsageCategory || (UsageCategory = {}));
/** Shared kobo validation used by both debit and credit paths. */
function validateKobo(amountKobo) {
    if (!Number.isInteger(amountKobo) || amountKobo < 0) {
        throw new Error('Amount must be a positive integer in kobo');
    }
}
export class BillingLedger {
    db; // Type would be D1Database from @cloudflare/workers-types
    constructor(db) {
        this.db = db;
    }
    /**
     * Records a usage debit for a tenant.
     */
    async recordUsage(tenantId, category, amountKobo, description, metadata) {
        validateKobo(amountKobo);
        const entry = {
            id: crypto.randomUUID(),
            tenantId,
            type: LedgerEntryType.DEBIT,
            category,
            amountKobo,
            description,
            ...(metadata !== undefined ? { metadata } : {}),
            createdAt: new Date(),
        };
        // In a real implementation, this would insert into D1:
        // await this.db.prepare('INSERT INTO ledger_entries ...').bind(...).run();
        return entry;
    }
    /**
     * Records a credit for a tenant (wallet top-up, refund, promotional credit,
     * or subscription payment).
     */
    async recordCredit(tenantId, category, amountKobo, description, metadata) {
        validateKobo(amountKobo);
        const entry = {
            id: crypto.randomUUID(),
            tenantId,
            type: LedgerEntryType.CREDIT,
            category,
            amountKobo,
            description,
            ...(metadata !== undefined ? { metadata } : {}),
            createdAt: new Date(),
        };
        // In a real implementation, this would insert into D1:
        // await this.db.prepare('INSERT INTO ledger_entries ...').bind(...).run();
        return entry;
    }
    /**
     * Calculates the current balance for a tenant.
     *
     * @stub This implementation always returns 0. Real balance calculation
     * requires D1 database wiring (separate task). Any caller that gates
     * actions on this value will silently allow unlimited usage until
     * D1 integration is complete.
     */
    async getTenantBalance(tenantId) {
        logger.warn('getTenantBalance is a stub and always returns 0. D1 integration is required for real balance calculation.', { tenantId });
        // Real implementation:
        // const result = await this.db.prepare(
        //   'SELECT SUM(CASE WHEN type = "CREDIT" THEN amountKobo ELSE -amountKobo END) as balance ' +
        //   'FROM ledger_entries WHERE tenantId = ? AND deletedAt IS NULL'
        // ).bind(tenantId).first();
        // return result.balance ?? 0;
        return 0;
    }
}
