// @webwaka/core/dist/core/document/index.js
/**
 * CORE-11: Document & Contract Management
 * Blueprint Reference: Part 10.5 (Real Estate), Part 10.12 (Legal)
 *
 * Secure system for generating, signing, and storing legal documents.
 *
 * Tenant Isolation: every mutating and querying method requires a tenantId.
 * Documents are scoped per tenant — cross-tenant leakage is impossible by construction.
 */
export class DocumentEngine {
    documents = new Map();
    /**
     * Creates a new document from a template, scoped to the tenant.
     */
    createDocument(tenantId, title, content) {
        const doc = {
            id: `doc_${crypto.randomUUID()}`,
            tenantId,
            title,
            content,
            status: 'draft',
            signatures: [],
            createdAt: new Date(),
        };
        this.documents.set(doc.id, doc);
        return doc;
    }
    /**
     * Requests signatures for a document, scoped to the tenant.
     */
    requestSignatures(tenantId, documentId) {
        const doc = this.documents.get(documentId);
        if (!doc || doc.tenantId !== tenantId)
            throw new Error('Document not found');
        if (doc.status !== 'draft') {
            throw new Error('Document must be in draft status to request signatures');
        }
        doc.status = 'pending_signature';
        return doc;
    }
    /**
     * Signs a document, scoped to the tenant.
     */
    signDocument(tenantId, documentId, userId, ipAddress) {
        const doc = this.documents.get(documentId);
        if (!doc || doc.tenantId !== tenantId)
            throw new Error('Document not found');
        if (doc.status !== 'pending_signature') {
            throw new Error('Document is not pending signature');
        }
        if (doc.signatures.some(s => s.userId === userId)) {
            throw new Error('User has already signed this document');
        }
        const signature = {
            userId,
            timestamp: new Date(),
            ipAddress,
            hash: this.generateSignatureHash(doc.content, userId, ipAddress),
        };
        doc.signatures.push(signature);
        doc.status = 'signed';
        return doc;
    }
    generateSignatureHash(content, userId, ipAddress) {
        return `hash_${userId}_${Date.now()}`;
    }
}
