// @webwaka/core/dist/core/events/index.js
/**
 * CORE-14: Event Bus Primitives
 * Blueprint Reference: Part 2 (Platform Architecture — "Event-Driven: NO direct inter-DB access")
 *
 * Canonical event contracts for the entire WebWaka OS v4 platform.
 * Every domain emits typed DomainEvents; consumers subscribe by event type.
 *
 * This module defines shapes, constants, and the emitEvent utility.
 * Actual queue/bus wiring (Cloudflare Queues) is a separate concern.
 */
/**
 * Well-known event type names for all WebWaka OS v4 domains.
 *
 * Consumers MUST use these constants — never raw string literals — so that
 * renaming an event causes a compile-time error instead of a silent mismatch.
 */
export var WebWakaEventType;
(function (WebWakaEventType) {
    // ─── Auth ──────────────────────────────────────────────────────────────────
    WebWakaEventType["AUTH_USER_LOGIN"] = "auth.user.login";
    WebWakaEventType["AUTH_USER_LOGOUT"] = "auth.user.logout";
    WebWakaEventType["AUTH_TOKEN_REFRESHED"] = "auth.token.refreshed";
    // ─── KYC ───────────────────────────────────────────────────────────────────
    WebWakaEventType["KYC_SUBMITTED"] = "kyc.submitted";
    WebWakaEventType["KYC_VERIFIED"] = "kyc.verified";
    WebWakaEventType["KYC_REJECTED"] = "kyc.rejected";
    // ─── Booking ───────────────────────────────────────────────────────────────
    WebWakaEventType["BOOKING_CONFIRMED"] = "booking.confirmed";
    WebWakaEventType["BOOKING_CANCELLED"] = "booking.cancelled";
    // ─── Chat ──────────────────────────────────────────────────────────────────
    WebWakaEventType["CHAT_MESSAGE_SENT"] = "chat.message.sent";
    WebWakaEventType["CHAT_CHANNEL_CREATED"] = "chat.channel.created";
    // ─── Document ──────────────────────────────────────────────────────────────
    WebWakaEventType["DOCUMENT_CREATED"] = "document.created";
    WebWakaEventType["DOCUMENT_SIGNED"] = "document.signed";
    // ─── Billing ───────────────────────────────────────────────────────────────
    WebWakaEventType["BILLING_DEBIT_RECORDED"] = "billing.debit.recorded";
    WebWakaEventType["BILLING_CREDIT_RECORDED"] = "billing.credit.recorded";
    // ─── Notification ──────────────────────────────────────────────────────────
    WebWakaEventType["NOTIFICATION_SENT"] = "notification.sent";
    WebWakaEventType["NOTIFICATION_FAILED"] = "notification.failed";
})(WebWakaEventType || (WebWakaEventType = {}));
/**
 * Factory that creates a well-formed DomainEvent with a generated id and
 * current timestamp.
 *
 * @param type      A WebWakaEventType constant.
 * @param tenantId  Tenant that owns the event.
 * @param payload   Domain-specific payload.
 */
export function createEvent(type, tenantId, payload) {
    return {
        id: crypto.randomUUID(),
        type,
        tenantId,
        occurredAt: new Date(),
        payload,
    };
}
/**
 * emitEvent — Publish a domain event to the platform event bus.
 *
 * Strategy:
 *   1. Write to EVENTS KV namespace as an outbox entry (TTL 24h)
 *   2. If EVENT_BUS_URL is set, also forward via HTTP POST (fire-and-forget)
 *
 * This function NEVER throws — failures are logged and swallowed so that
 * the calling business logic is never blocked by event bus unavailability.
 *
 * @param env       Worker environment bindings (must include EVENTS KV)
 * @param eventType Canonical event type string (e.g. "civic.member.created")
 * @param tenantId  Tenant that owns this event
 * @param payload   Domain-specific payload (must be JSON-serialisable)
 */
export async function emitEvent(env, eventType, tenantId, payload) {
    const event = {
        id: crypto.randomUUID(),
        type: eventType,
        tenantId,
        occurredAt: new Date().toISOString(),
        payload,
    };
    const key = `event:${Date.now()}:${crypto.randomUUID()}`;
    const body = JSON.stringify(event);
    // 1. Write to KV outbox
    if (env.EVENTS) {
        try {
            await env.EVENTS.put(key, body, { expirationTtl: 86400 });
        }
        catch {
            // Non-fatal — continue to HTTP delivery
        }
    }
    // 2. HTTP delivery (fire-and-forget)
    if (env.EVENT_BUS_URL) {
        try {
            await fetch(env.EVENT_BUS_URL, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body,
                signal: AbortSignal.timeout(5000),
            });
        }
        catch {
            // Non-fatal — event is already in KV outbox
        }
    }
}
